src/algebra/ordered_monoid.lean

```diff
@@ -396,6 +396,48 @@ instance [add_comm_monoid α] : add_comm_monoid (with_top α) :=
   ..@additive.add_comm_monoid _ $ @comm_monoid_with_zero.to_comm_monoid _ $
     @with_zero.comm_monoid_with_zero (multiplicative α) _ }
 
+section sub_neg
+
+instance [has_neg α] : has_neg (with_top α) :=
+⟨option.map (λ (a : α), -a)⟩
+
+@[simp] lemma neg_top [has_neg α] : - (⊤ : with_top α) = ⊤ := rfl
+lemma neg_coe [has_neg α] (x : α) : (-x : with_top α) = ((-x : α) : with_top α) := rfl
+lemma coe_neg [has_neg α] (x : α) : ((-x : α) : with_top α) = -x := rfl
+@[simp] lemma neg_eq_top [has_neg α] (x : with_top α) : - x = ⊤ ↔ x = ⊤ :=
+by { induction x using with_top.rec_top_coe; simp [neg_coe] }
+
+/-- A `with_top α` can have a subtraction if `[has_sub α]`. However, this might
+not be the subtraction that is preferred, for example `ennreal.has_sub`. -/
+instance [has_sub α] : has_sub (with_top α) :=
+⟨λ o₁ o₂, o₁.bind (λ a, o₂.map (λ b, a - b))⟩
```

Comment:
Maybe I'm misunderstanding what's going on here, but this comment seems to be shouting: "This should not be an instance!" Can you just use `local attribute [instance]` where needed (or, perhaps, abandon this definition entirely)?